jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java

```diff
@@ -592,6 +596,7 @@ public ClientSelectorManager(Executor executor, Scheduler scheduler, int selecto
         protected EndPoint newEndPoint(SelectableChannel channel, ManagedSelector selector, SelectionKey selectionKey)
         {
             EndPoint endPoint = ClientConnector.this.newEndPoint(channel, selector, selectionKey);
+            ClientConnector.this.getBeans(ClientConnector.ConnectListener.class).forEach(listener -> listener.onConnectSuccess((SocketChannel)channel));
```

Comment:
Move the "success" event just after connection.
For blocking should be just after the `socket().connect(...)` call.
I'll leave to you finding the similar place for non-blocking.

Reply:
The best place for Success event for the non-blocking connection is `ClientSelectorManager.connectionOpened` event, but in that case `SocketChanel` is missed.
So I would like to keep "success" event in the current place, as `newEndPoint` event indicates registration of a channel via `connect()`.

Reply:
@arsenalzp I would like the "success" event to be emitted as soon as possible.

Method `newEndPoint()` is called only after dispatching the `EndPoint` creation to a new thread, so the creation task may stay in the executor queue for a while.

The correct place is `ManagedSelector.processConnect()`, please have a look there.

For blocking, it is just after `socketChannel.socket().connect()`.

Reply:
Because `ManagedSelector.processConnect()` is `private` you can add a new overridable method `SelectorManager.onConnected()`, so that you can wire the "success" event for both the blocking and non-blocking case in the same place.